### ISSUE TITLE
code-complexity-control: calculateFileComplexity() counts split('\n').length, so every newline-terminated file measures one line too long and a file at exactly maxFileLines is rejected

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#127]
**Version:** 7.21.0
**Law:** `code-complexity-control` — `CodeComplexityControlLaw.calculateFileComplexity()` in `dist/checkers/code-quality-laws/code-complexity-control.js`

Kind: it reports code that is correct — a false positive on every file that sits at exactly `maxFileLines`, and a wrong line number on every file it reports at all.

## What happened

A file trimmed down until ESLint was satisfied still failed the commit hook:

```
❌ FAILED CONSTITUTIONAL LAWS:

1. Code Complexity Control [1 violations] (Score: 90/100)
   • File too long: libs/shell/feature/src/lib/activate/activate.page.ts (301 lines, max 300)

   💡 What the law is asking for:
   • Break down large files into smaller, focused modules
   • Extract reusable functions and components
   • Large number of dependencies detected (60 > 50) - consider dependency consolidation
```

At that moment the file held **300** lines:

```
$ wc -l libs/shell/feature/src/lib/activate/activate.page.ts
300
```

and ESLint — configured in the same repo to measure the same thing, `'max-lines': ['error', { max: 300, skipBlankLines: false, skipComments: false }]` — ran on that exact tree and reported nothing:

```
$ npx nx run-many -t test lint -p shell-feature --configuration=ci --skip-nx-cache
...
Successfully ran targets test, lint for project shell-feature   (exit 0)
```

Three tools, one file, at the boundary: `wc -l` says 300, ESLint says 300 and passes, RoC says 301 and blocks the commit. Deleting one more line (299 by `wc -l`) let the commit through — the same off-by-one seen from the other side.

## What is wrong

`calculateFileComplexity()`, line 120:

```js
static calculateFileComplexity(content) {
    const lines = content.split('\n').length;   // ← counts the empty string after the final newline
```

consumed at line 75:

```js
const maxFileLines = config.thresholds?.codeQuality?.maxFileLines ?? 300;
if (complexity.lines > maxFileLines) {
    violations.push(`File too long: ${...} (${complexity.lines} lines, max ${maxFileLines})`);
}
```

Every POSIX-conformant text file ends with a newline, so `split('\n')` returns a trailing `""` that is counted as a line. The comparison is `>`, so the practical effect is that the real limit is `maxFileLines - 1`, and the number printed in the violation is one higher than anything the author can see in their editor.

## Reproduction

Called through the law's own exported static method, not inferred from the audit output:

```js
const { CodeComplexityControlLaw } = require('ruleofcode/dist/checkers/code-quality-laws/code-complexity-control.js');
const mk = (n) => Array.from({ length: n }, (_, i) => '// line ' + (i + 1)).join('\n') + '\n';

for (const n of [1, 299, 300, 301]) {
  const lines = CodeComplexityControlLaw.calculateFileComplexity(mk(n)).lines;
  console.log(`${n}-line file -> law.lines = ${lines}   (> 300 ? ${lines > 300})`);
}
```

```
1-line file   -> law.lines = 2     (> 300 ? false)
299-line file -> law.lines = 300   (> 300 ? false)
300-line file -> law.lines = 301   (> 300 ? true)     ← rejected at exactly the limit
301-line file -> law.lines = 302   (> 300 ? true)
```

And against the real file, on the current `develop` of the same repo:

```
libs/shell/feature/src/lib/activate/activate.page.ts
  wc -l     = 296
  law.lines = 297
```

## What it cost me

The commit hook is a ~75s full audit, so each guess costs a run. The reported number matches nothing the author can see, which is the expensive part: the natural reading of "301 lines, max 300" on a file whose editor says 300 is "I must have miscounted", not "the counter is off by one". I moved a constant into a separate module, re-ran, still failed, trimmed a doc comment, re-ran, still failed — and only then read `dist/` and found the phantom line.

The knock-on is worse than the delay. This law and the project's own ESLint `max-lines` are configured to the same number for the same reason, and they disagree by one, so a file can be green in `npm run verify` and red at commit time with no way to tell from either message which one is right. Two files in that directory now sit at 296–298 lines, i.e. two lines from a ceiling that is not where either tool says it is.

## Is the same logic somewhere else?

Yes, once, and it is the same mistake:

- `dist/checkers/angular-laws/defer-heavy-blocks.js:16` — `const lines = html.split('\n').length;` compared to `deferHeavyBlocksMinLines` (default 250). Lower stakes: it is a soft heuristic scoped to `*-dashboard.component.html`, not a hard gate. Same fix.

Not affected, checked rather than assumed: the many `content.slice(0, index).split('\n').length` calls across the Python and FE-architecture checkers (`requests-have-timeout`, `no-shell-injection`, `no-mutable-default-arguments`, `fastapi-typed-responses`, `poll-via-sanctioned-source`, `defensive-array-coercion`, `no-risk-literals`, `no-passthrough-cast`, `insecure-deserialization`, `strict-expected-failures`) are computing a **1-based line number for a match offset**, where that idiom is correct.

## Suggested fix

1. Count lines of text rather than split segments, in `calculateFileComplexity()`:
   ```js
   const lines = content.length === 0
       ? 0
       : content.split('\n').length - (content.endsWith('\n') ? 1 : 0);
   ```
2. Same change at `defer-heavy-blocks.js:16`.
3. Worth a regression test at the boundary specifically — a fixture of exactly `maxFileLines` lines must pass and `maxFileLines + 1` must fail. An off-by-one is invisible to any test that only checks a clearly-too-long file.

Measured behaviour of the proposal against the current code:

| content | `wc -l` | current | proposed |
| --- | --- | --- | --- |
| `""` (empty) | 0 | **1** | 0 |
| `"a\nb\nc\n"` | 3 | **4** | 3 |
| `"a\nb\nc"` (no trailing newline) | 2 | 3 | 3 |

The last row is the one place the proposal deliberately does not match `wc -l`: `wc` counts terminators, so it sees two, while the file plainly holds three lines of text. Counting the text is what a lines-of-code limit is about — and it is what the current code already gets right by accident. The defect is only the phantom line on the first two rows, which is every normal file.

## What I did NOT verify

- Whether this reproduces on 7.22.0 — I am on 7.21.0, pinned in this project's `package.json`, and have not re-staged on the latest.
- CRLF checkouts. The affected file is LF (`content.includes('\r\n') === false`, checked). Reading the code, `split('\n')` should count `\r\n` files identically since each segment merely keeps a trailing `\r`, but I have not run the repro on a CRLF checkout to confirm it.
- Whether any other checker holds a differently-named whole-file line count with the same mistake. I grepped for the literal pattern `split('\n').length` across `dist/` and classified every hit; I did not look for other spellings (`split(/\n/)`, a shared helper, etc.).
- Whether `maxFileLines` behaves differently when set explicitly — this project does not set it, so the default `?? 300` is what ran.

## My config

`ruleofcode.config.json`, `thresholds.codeQuality` — note `maxFileLines` is absent, so the `?? 300` default applies:

```json
{
  "maxComplexity": 20,
  "minFunctionComplexity": 10,
  "maxDependencies": 50,
  "minDuplicationBlockSize": 250,
  "minCommentRatioPercent": 10,
  "minFileSizeForComments": 50
}
```

**Comment by ruleofcodedev** [comment]
Confirming this independently on `ruleofcode@7.21.0`, and adding one consequence the report does not draw out: **with a formatter in the repo there is no workaround available to the author.**

The report's repro removes a *line*. I removed only the *terminator*, leaving the content byte-identical otherwise:

```
probe file, 300 lines of `// filler NNN`, POSIX-terminated
  wc -l                       300
  split('\n').length          301
  ESLint max-lines:300        exit 0    (skipBlankLines:false, skipComments:false)
  RoC  maxFileLines:300       exit 1    File too long: ... (301 lines, max 300)

same file, final \n stripped, nothing else touched
  wc -l                       299
  split('\n').length          300
  RoC                         passes — the violation disappears
```

So the check is satisfied by making the file non-POSIX. That matters because the trailing newline is not the author's choice: `prettier` writes it unconditionally, and `prettier --check` is in the same `verify` pipeline as `roc` in this repo. The two gates therefore disagree about a file at exactly the limit, and each one's fix breaks the other:

```
npm run verify
  ... && npm run format:check && ... && npm run roc
```

An author at exactly `maxFileLines` cannot satisfy both by editing whitespace — only by deleting a real line of source. That is why the effective limit is `maxFileLines - 1` for every formatted repo, rather than an off-by-one that shows up only at the boundary.

The row in your table for `"a\nb\nc"` already anticipates this case; I am only noting that in practice that state is unreachable, so the proposed fix is the only way out.

Verified on the same file the report cites — `activate.page.ts` in an Angular/Nx repo — and again on a synthetic 300-line probe, both with `--skip-nx-cache` / a fresh audit.

**Comment by ruleofcodedev** [comment]
Fixed in #134, merged to `develop`. Ships in the next release.

`countLines()` subtracts the terminator, and an empty file now measures 0 rather than 1.

Worth saying plainly: the number this law printed was wrong for **every** file it named, not only at the boundary. A developer who checked with `wc -l` got a different answer every time, which is why the finding read as a bug in the tool. It was.

The regression test asserts the reported **count**, not merely the presence of a finding — a fix that made 301 pass while still printing the wrong number would have been worse than the bug.
